i18n-helpers/src/lib.rs

```diff
@@ -12,112 +12,487 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use once_cell::sync::Lazy;
-use regex::Regex;
+use mdbook::utils::new_cmark_parser;
+use pulldown_cmark::{Event, Tag};
+use std::ops::Range;
 
-static PARAGRAPH_SEPARATOR: Lazy<Regex> = Lazy::new(|| Regex::new(r"\n\n+").unwrap());
+/// A translatable message.
+#[derive(PartialEq, Debug)]
+pub struct Message {
+    /// Line number where this message begins.
+    line: usize,
 
-/// Extract paragraphs from text.
-///
-/// Paragraphs are separated by at least two newlines. Returns an
-/// iterator over line numbers (starting from 1) and paragraphs.
-pub fn extract_paragraphs(text: &str) -> impl Iterator<Item = (usize, &str)> {
-    // TODO: This could be made more sophisticated by parsing the
-    // Markdown and stripping off the markup characters.
-    //
-    // As an example, a header like "## My heading" could become just
-    // "My heading" in the `.pot` file. Similarly, paragraphs could be
-    // unfolded and list items could be translated one-by-one.
-
-    // Skip over leading empty lines.
-    let trimmed = text.trim_start_matches('\n');
-    let mut matches = PARAGRAPH_SEPARATOR.find_iter(trimmed);
-    let mut lineno = 1 + text.len() - trimmed.len();
-    let mut last = 0;
-
-    std::iter::from_fn(move || match matches.next() {
-        Some(m) => {
-            let result = (lineno, &trimmed[last..m.start()]);
-            lineno += trimmed[last..m.end()].lines().count();
-            last = m.end();
-            Some(result)
+    /// Span of the input text containing this message.
+    span: Range<usize>,
+}
+
+impl Message {
+    fn new(line: usize, span: Range<usize>) -> Self {
+        Self { line, span }
+    }
+
+    /// Get the text of this message, as a slice of the document from which it was generated.
+    pub fn text<'doc>(&self, document: &'doc str) -> &'doc str {
+        &document[self.span.start..self.span.end]
```

Comment:
I suppose this could be simplified to just

```suggestion
        &document[self.span]
```

Reply:
With a `.clone()`, yes.